specs/capella/beacon-chain.md

```diff
@@ -371,15 +272,29 @@ def process_block(state: BeaconState, block: BeaconBlock) -> None:
 
 ```python
 def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
-    num_withdrawals = min(MAX_WITHDRAWALS_PER_PAYLOAD, len(state.withdrawal_queue))
-    dequeued_withdrawals = state.withdrawal_queue[:num_withdrawals]
-
-    assert len(dequeued_withdrawals) == len(payload.withdrawals)
-    for dequeued_withdrawal, withdrawal in zip(dequeued_withdrawals, payload.withdrawals):
-        assert dequeued_withdrawal == withdrawal
-
-    # Remove dequeued withdrawals from state
-    state.withdrawal_queue = state.withdrawal_queue[num_withdrawals:]
+    epoch = get_current_epoch(state)
+    index = state.last_withdrawal_validator_index
+    for withdrawal in payload.withdrawals:
+        index = ValidatorIndex((index + 1) % len(state.validators))
+        while index != withdrawal.validator_index:
+            val state.validators[index]
+            balance = state.balances[index]
+            assert !is_fully_withdrawable_validator(val, balance, epoch):
+            assert !is_partially_withdrawable_validator(val, balance):
+            index += ValidatorIndex((index + 1) % len(state.validators))
+        assert withdrawal.validator_index == index
+        val = state.validators[index]
+        balance = state.balances[index]
+        fully_withdrawable = is_fully_withdrawable_validator(val, balance, epoch)
+        partial_withdrawable = is_partially_withdrawable_validator(val, balance)
+        assert fully_withdrawable || partial_withdrawable
+        if fully_withdrawable:
+            assert withdrawal.amount == balance
+            decrease_balance(state, index, balance)
+        else:
+            assert withdrawal.amount == balance - MAX_EFFECTIVE_BALANCE
```

Comment:
you'd also want to check that the `withdrawal.address` matches the validator's withdrawal credentials